--- e2fsprogs-1.42.6.orig/Makefile.in	2012-07-06 19:07:27.000000000 +0530
+++ e2fsprogs-1.42.6/Makefile.in	2012-10-09 15:18:43.000000000 +0530
@@ -61,7 +61,7 @@
 	-test -d doc && cd doc && $(MAKE) distclean
 
 install: subs all-libs-recursive install-progs-recursive \
-  install-shlibs-libs-recursive install-doc-libs
+  install-shlibs-libs-recursive 
 	if test ! -d e2fsck && test ! -d debugfs && test ! -d misc && test ! -d ext2ed ; then $(MAKE) install-libs ; fi
 
 install-strip: subs all-libs-recursive install-strip-progs-recursive \
